### DISCUSSION TITLE
Datasets

### DISCUSSION THREAD

**Discussion by Mayitzin** [Mayitzin/ahrs#20] [General]
A great number of datasets can be found online, and I'm permanently looking for good ones. I'd recommend to use the ones including:

- Gyroscope data.
- Accelerometer data.
- Magnetometer data.
- Ground truth of orientations.

Some may lack the data of one of the sensors. Some others might have even more data than you expect (like videos or images for visual dometry.)

Check if you can use them to test your applications. Here I will post some of the ones I've found useful, their pros and cons. Maybe you can get some nice results with them.

**Comment by Mayitzin** [comment]
RepoIMU
---

Location: https://github.com/agnieszkaszczesna/RepoIMU
Author: Agnieszka Szczesna, Przemysław Skurowski, Przemysław Pruszowski, Damian Pęszor, Marcin Paszkuta, and Konrad Wojciechowski.
Documentation: https://www.researchgate.net/publication/307076277

Is the first dataset that comes to mind. It has full 9-dof recordings at 100 Hz, synchronized to their ground truths as quaternions. One of the files of this dataset was actually used for initial tests of `AHRS`, and you might find it in the `tests` folder.

**Pros:**
- Full data of the 3 types of main sensors.
- Ground truth (as quaternions).
- Sensor data and ground truth are already synchronized at 100 Hz.
- Includes static recordings to analyze stabilty of sensor signals.
- Single CSV File per recording with full data.

**Cons:**
- Recordings do not reflect real user applications, only controlled angular motions.
- Does not include translation (position of sensor frame w.r.t. the global frame.)

I personally would use this dataset to understand the angular motion, and test drift after angular integration.

**Reply by anisghaoui** [reply]
I just tried this experiment and yes and I do obtain the same result that validate the estimation on the X axis by a madgwick filter.

In this figure you can see the error of estimation of madgwick vs vicon 
![vicon_madgwick](https://user-images.githubusercontent.com/37940201/106897186-e250d880-66f2-11eb-8dd1-c49be3eb5924.png)


Did you test Madgwick on other experiments where axes other than X axis are put into exercise?

**Reply by anisghaoui** [reply]
To a further extent, i swapped X and Z axes in gyroscopes, accelerometers and vicon orientation (after changing it to angles) as follows
```python
   # swapping Z with X
    imu_gyro[:, 0], imu_gyro[:, 2] = imu_gyro[:,2], imu_gyro[:,0].copy()
    imu_acce[:, 0], imu_acce[:, 2] = imu_acce[:,2], imu_acce[:,0].copy()
    # convert vicon to angles
    vicon_ori = QuaternionArray(vicon_quater).to_angles()
    vicon_ori[:, 0], vicon_ori[:, 2] = vicon_ori[:, 2], vicon_ori[:, 0].copy()
```
it is expected to obtain the same result as the previous graphic but on the z axis, isn't it?
Well, no :
![Madgwick](https://user-images.githubusercontent.com/37940201/106895244-46be6880-66f0-11eb-8a5c-504deb0ca3d2.png)

![vicon](https://user-images.githubusercontent.com/37940201/106895251-4756ff00-66f0-11eb-8d14-763335675f13.png)

![madgwick_vs_vicon](https://user-images.githubusercontent.com/37940201/106897330-11674a00-66f3-11eb-9886-516ca993182e.png)

You can see that the result is far from being correct.

Edit : the same issue appears when swapping X with Y.

**Reply by Mayitzin** [reply]
The RepoIMU dataset offers the reference values as quaternions only. The roll-pitch-yaw angles and other representations are, therefore, computed from these quaternions. Because the given references are quaternions, it is first recommended to test with this representation first.

All reference orientations have been rotated so that their initial orientation is equal to the identity quaternion `[1, 0, 0, 0]`. The estimated orientations have to be rotated to match their corresponding references, OR the reference orientations must be rotated to match the estimated orientations. According to the original article of the dataset:

> When using IMU sensor measurements provided with this paper ... it is necessary to cancel the rotation between coordinate systems ... First, we need to compute momentary differences of IMU (Q<sub>i</sub>) and reference (Q<sup>ref</sup><sub>i</sub> ) orientation quaternions as <img src="https://render.githubusercontent.com/render/math?math=\Delta Q_i = Q_i \cdot conj(Q_i^{ref})">, and then compute the mean difference <img src="https://render.githubusercontent.com/render/math?math=\bar{\Delta Q_i}">. Next we need to rotate quaternions from IMU coordinates to the reference with: <img src="https://render.githubusercontent.com/render/math?math=Q^{rot}_i=conj(\bar{\Delta Q_i})\cdot Q_i">

But the recordings have the big advantage of being immobile during the first samples. Therefore, we can use the conjugate of the first reference orientation to rotate ALL estimated quaternions. Remember the conjugate of a quaternion represents the "inverse rotation". Performing the [Hamilton product](https://en.wikipedia.org/wiki/Quaternion#Hamilton_product) between the conjugated first reference quaternion and all estimated quaternions will align the latter to the reference frame.

It can be done with:

```python
def rotate_quaternions(q, Q):
    return np.c_[
        q[0]*Q[:, 0] - q[1]*Q[:, 1] - q[2]*Q[:, 2] - q[3]*Q[:, 3],
        q[0]*Q[:, 1] + q[1]*Q[:, 0] + q[2]*Q[:, 3] - q[3]*Q[:, 2],
        q[0]*Q[:, 2] - q[1]*Q[:, 3] + q[2]*Q[:, 0] + q[3]*Q[:, 1],
        q[0]*Q[:, 3] + q[1]*Q[:, 2] - q[2]*Q[:, 1] + q[3]*Q[:, 0]]
```

where `q` is a single quaternion and `Q` is the array of quaternions to rotate about `q`. Then, we can compare the references against the estimations with the **Quaternion Angle Difference** included in the `metrics` submodule.

For this test I use a file from the session "11" of the "TStick", which is a free motion about the three axes. This would help us to further test the estimation in a "less predictable" environment. This type of motion is, I think, what you would like to be testing.

```python
import numpy as np
import ahrs
data = np.genfromtxt("./TStick/TStick_Test11_Trial1.csv", dtype=float, delimiter=';', skip_header=2)
ref_quaternions = data[:, 1:5]
acc_data = data[:, 5:8]
gyr_data = data[:, 8:11]
orientation = ahrs.filters.Madgwick(gyr=gyr_data, acc=acc_data)
ref_q0_conjugate = ref_quaternions[0]*np.array([1.0, -1.0, -1.0, -1.0])
rotated_qts = rotate_quaternions(ref_q0_conjugate, orientation.Q)
errors = ahrs.utils.metrics.qad(ref_quaternions, rotated_qts)
```

The errors estimated by `qad` are returned as an array with a float per comparison between quaternions. Plotting all these values will seem like:

![grafik](https://user-images.githubusercontent.com/4802163/107293956-d35c8400-6a6c-11eb-912b-a1b8e080a9f7.png)

These quaternoins can be represented as angles within the range [-pi, pi], and they can be compared with a simple euclidean distance:

```python
ref_angles = ahrs.QuaternionArray(ref_quaternions).to_angles()
angles = ahrs.QuaternionArray(rotated_qts).to_angles()
errors = ahrs.utils.metrics.euclidean(ref_angles, angles, axis=1)
```

Plotting these results you see something like:

![grafik](https://user-images.githubusercontent.com/4802163/107296583-f89fc100-6a71-11eb-9a5a-eadaa71f09d1.png)

There are noticeable jumps when the angles approach, and cross, the threshold above pi. When the estimated angle is "jumping", but its reference is not, it would give you an error close to 2pi (per axis "crossing"), because the simple Euclidean distance does not consider that these angles are actually quite close.

This effect can also be seen in the Y- and Z-axis errors of the first plot you've added. The errors jump between 0 and 2pi, or values close to them, while the X-axis error stays around 2pi (which is actually 0 in geometric terms). In this example they are the spikes in the computed errors.

It is, therefore, much more preferred to compare the estimated orientations using the Quaternions, which we have at hand.

Just remember to rotate your estimations when using the RepoIMU dataset, to represent the estimations in the reference frame.

P.D. Now planning to add the rotation function to the class `QuaternionArray`, so that users can rotate all quaternions by any given quaternion.

**Reply by anisghaoui** [reply]
Hi, thank you for the answer. 
I totally agree with you on quaternion comparison. I reproduced the same test and obtain the exact same results.
However, I still can't get my head around how the madgwick reference system. it is local to the sensor as he shows it in figure 6 of his [article](https://ieeexplore.ieee.org/abstract/document/5975346?casa_token=oZzjE7I6bEMAAAAA:qg7DvjEHTfvQarfN3knCM2cSsyvpT2fHicnzHUhFYMVuqcMuzMP3_YP9-TZoRUGbBZrPu6Vi55gwOcE).

I am trying to perform unitary tests where I only rotate a single axis of my phone  with a **gain of 0**:
- phone is on table, face towards my left. turned 180° turn in the positive direction

  - gyro integration
![rot_x_integ](https://user-images.githubusercontent.com/37940201/107350859-3cbeb000-6aca-11eb-8483-7b43ae90aba9.png)
  - madgwick estimation
![rot_x_madg](https://user-images.githubusercontent.com/37940201/107351628-1fd6ac80-6acb-11eb-8a0b-7e35796b7665.png)


- phone is in my hand held against the table, front camera facing me. Span it 180° positive
  - gyro integration
![rot_y_integ](https://user-images.githubusercontent.com/37940201/107350864-3d574680-6aca-11eb-8e2f-3661cd9893bb.png)
  - madgwick estimation
![rot_y_madg](https://user-images.githubusercontent.com/37940201/107350866-3defdd00-6aca-11eb-8f86-74ebacc995f5.png)

- phone is on table, front camera face to the ceiling  and span 360° positive

  - gyro integration
![rot_z_integ](https://user-images.githubusercontent.com/37940201/107350867-3e887380-6aca-11eb-97c0-74d22d87c1fc.png)

  - madgwick estimation
![rot_z_madg](https://user-images.githubusercontent.com/37940201/107350869-3e887380-6aca-11eb-96bf-71983dc5173e.png)

While both X- and Z- axes seem correct, the graph on Y-axis rotation seems to be wrong.
I don't have any intuition on why that is happening. Do you happen to have seen this problem before?

**Reply by anisghaoui** [reply]
Hi, I am back at it.

I just tried to run a test with sequence `RepoIMU/TStick/TStick_Test03_Trial1.csv`. and I obtain the same result as my unit test. 
The trial in the ground truth dataset consists into spinning around the axis 1 negative turn then 1 positive turn. let's integrate the gyro to check that:
![integ_gyro](https://user-images.githubusercontent.com/37940201/107933756-763a6400-6f7f-11eb-899f-0a986b61bd23.png)
Indeed, it is.
So let's compare the reference quaternion to Madgwick's:
![ref_quat](https://user-images.githubusercontent.com/37940201/107933830-8e11e800-6f7f-11eb-84d9-ede53471aa68.png)
![estim_quat](https://user-images.githubusercontent.com/37940201/107933826-8d795180-6f7f-11eb-8494-db7e89d57511.png)

The are visually very similar. Let's convert both to Roll Yaw Pitch to check if it corresponds to reality:
![ref_ori](https://user-images.githubusercontent.com/37940201/107933938-b4378800-6f7f-11eb-90f6-27b7cd360276.png)
![estim_ori](https://user-images.githubusercontent.com/37940201/107933937-b39ef180-6f7f-11eb-8cf6-dc0adb337948.png)

There seems to be a problem with the conversion. I am wondering if this isn't a **pole singularity problem**...
**We may need to recheck the `to_angles()` conversion**.

The code from this [website](http://www.euclideanspace.com/maths/geometry/rotations/conversions/quaternionToEuler/) translates to:
```python
def to_angles(q):
    test = q[1] * q[2] + q[3] * q[0]
    if test > 0.499:  # singularity at north pole
        print("singularity at north pole")
        phi = 2 * np.arctan2(q[1], q[0])
        theta = M_PI / 2
        psi = 0
        return np.c_[phi, theta, psi]
    if test < -0.499:  # singularity at south pole
        print("singularity at south pole")
        phi = -2 * np.arctan2(q[1], q[0])
        theta = -  M_PI / 2
        psi = 0
        return np.c_[phi, theta, psi]

    phi = np.arctan2(2.0 * (q[0] * q[1] + q[2] * q[3]), 1.0 - 2.0 * (q[1] ** 2 + q[2] ** 2))
    theta = np.arcsin(2.0 * (q[0] * q[2] - q[3] * q[1]))
    psi = np.arctan2(2.0 * (q[0] * q[3] + q[1] * q[2]), 1.0 - 2.0 * (q[2] ** 2 + q[3] ** 2))
    return np.c_[phi, theta, psi]
```
need to be made more Pythonesque. i tried it and no it doesn't solve the problem.

**Reply by Mayitzin** [reply]
Hi, I actually don't see a problem with the comparisons.

Do you refer to the similarity between the converted angles? Because that is expected. If you plot each axis against each other (reconstructed Yaw over reference Yaw, etc.) you will see that they are pretty similar, except for the jumps when they reach to -2pi or 2pi.

However, that is also expected. Although -2pi and 2pi are numerically far apart, they are very close geometrically. They basically represent the same angle.

**Reply by anisghaoui** [reply]
When you see the gyroscope integration, you can see that the Y-axis spins 1 negative turn then 1 positive turn. When you see the estimated and reference orientation, the Yaw simply doesn't follow that. I agree reference Yaw and reconstructed Yaw are similar because their quaternions are very very close. the problem isn't about +-2pi. The problem comes that the `to_angles()`conversion doesn't get you the real movement.

**Reply by Mayitzin** [reply]
I moved this topic to issue #31 because I think it is more convenient to discuss it there, as this discussion was created precisely to be about the RepoIMU dataset.

For further discussion about the transformation of quaternion to roll-pitch-yaew angles, please go there.

**Reply by PeterBorisenko** [reply]
My personal observation is that this dataset seems to provide deviated magnetometer data. 
The best results I could get was with simple gyroscope integration. Any use of magnetometer data makes a mess.

**Reply by PeterBorisenko** [reply]
I would like to know as well which variance for each sensor and which magnetic reference do you use with this data set in case of EKF?

**Comment by Mayitzin** [comment]
Attitude Estimation with Smartphones
---

- Location: https://github.com/tyrex-team/benchmarks-attitude-smartphones
- Author: Michel Thibaud, Hassen Fourati, Pierre Genevès, Nabil Layaïda
- Documentation: http://tyrex.inria.fr/mobile/benchmarks-attitude/
- License: [CeCILL](http://www.cecill.info/index.en.html)

This dataset is focused on smartphones, and their recordings are obtained from their embedded IMUs. Three smartphones with different sensor models are used:

| Model | Accelerometer | Gyroscope | Magnetometer |
|---|:--|:--|:--|
| iPhone 4S | STMicro STM33DH (100 Hz) | STMicro AGDI (100 Hz) | AKM 8975 (40 Hz) |
| iPhone 5 | STMicro LIS331DLH (100 Hz) | STMicro L3G4200D (100 Hz) | AKM 8963 (25 Hz) |
| Nexus 5 | InvenSense MPU6515 (200 Hz) | InvenSense MPU6515 (200 Hz) | AKM 8963 (60 Hz) |

Eight typical motions are recorded:

1. Augmented Reality.
2. Texting while walking.
3. Phoning while walking.
4. Swinging hand while walking.
5. In the front pocket while walking.
6. In the back pocket while walking.
7. In the front pocket while running.
8. In the hand while running.

**Pros:**
- Very practical use cases.
- Includes recordings with and without calibration of gyroscopes and magnetometers.
- Includes positions of sensor frame w.r.t. global frame.
- Ground truths are provided as **rotation matrices** and as **roll-pitch-yaw angles**.

**Cons:**
- Data is not synchronized.
- Some ground truth files have "jumps" with no data, due to the markers being covered during the recording.
- Files are in `MAT` format, which makes it difficult to read and organize.

**Comment by anisghaoui** [comment]
### OxIOD:  The  Dataset for Deep Inertial Odometry
----------
Location: http://deepio.cs.ox.ac.uk/
Authors: Changhao Chen, Peijun Zhao, Chris Xiaoxuan Lu, Wei Wang, Andrew Markham, Niki Trigoni
Date: 2018
Size: 1 GB
Duration: 14.72 hours total
Documentation: not found

This dataset is delivered with smartphone's IMU, Vicon and Google Tango phone (Visual-inertial SLAM). It was created to train their neural network.

Various phone positions and usages:

1. handbag      
2. handheld
3. pocket
4. trolley
5. running
6. slow walking
7. large scale
8. multi devices
9. multi users

Pros:

- Raw and synchronised data.
- All in .csv
- Both ground truths for small experiments and large scale
- frequency = 100 Hz

Cons:
- sync data have their timestamps overwritten as 1.5E11. You can still recover the correct timestamping.

**Comment by Mayitzin** [comment]
Berlin Robust Orientation Estimation Assessment Dataset (BROAD)
---

- Location: https://github.com/dlaidig/broad
- Authors: [D. Laidig](https://github.com/dlaidig), M. Caruso, A. Cereatti, T. Seel.
- Documentation: https://www.mdpi.com/2306-5729/6/7/72/pdf
- License: [Creative Commons BY 4.0](https://creativecommons.org/licenses/by/4.0/)

The dataset includes 3D inertial and magnetic data with synchronized optical marker-based ground truth measurement. It consists of 39 trials at different speeds and various types of movement. Thereof, 23 trials are performed in an undisturbed indoor environment, and 16 trials are recorded with deliberate magnetometer and accelerometer disturbances.

- IMU data recorded at 286 Hz using a nine-axis inertial sensor [myon aktos-t](https://www.myon.ch/aktos-t).
- Ground truth data at 120 Hz was obtained via an Optitrack OMC system consisting of eight [Flex13 cameras](https://optitrack.com/cameras/flex-13/).

Data in each file is:

- `imu_gyr` IMU gyroscope measurements in rad/s.
- `imu_acc` IMU accelerometer measurements in m/s2.
- `imu_mag` IMU magnetometer measurements in μT.
- `opt_quat` OMC ground truth orientation as a unit quaternion (w component first, ENU reference frame).
- `opt_pos` OMC ground truth position in m.
- `movement` Boolean array (0/1) indicating movement phases.
- `sampling_rate` sampling rate of the measurements in Hz (2000/7 Hz ≈ 286 Hz, HDF5: stored as an attribute).
- `info` short description of the file contents (HDF5: stored as an attribute).

**Pros:**
- Different environments and cases.
- Ground truths are provided as **quaternions** avoiding any gimbal lock problems and reducing the processed data.

**Cons:**
- Files are in either `HD5` or `MAT` format, which makes it difficult to read and organize. Extra tools are required to read the dataset (provided in its repository)
